// ngx-infinite-scroller/bundles/ngx-infinite-scroller.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('rxjs'), require('rxjs/operators'), require('is-number')) :
    typeof define === 'function' && define.amd ? define('ngx-infinite-scroller', ['exports', '@angular/core', '@angular/common', 'rxjs', 'rxjs/operators', 'is-number'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['ngx-infinite-scroller'] = {}, global.ng.core, global.ng.common, global.rxjs, global.rxjs.operators, global.isNumber));
}(this, (function (exports, core, common, rxjs, operators, isNumber) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var isNumber__default = /*#__PURE__*/_interopDefaultLegacy(isNumber);

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    ;
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }
    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var DirectiveStateService = /** @class */ (function () {
        function DirectiveStateService() {
        }
        Object.defineProperty(DirectiveStateService.prototype, "scrollTop", {
            get: function () {
                return this._el.nativeElement.scrollTop;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DirectiveStateService.prototype, "scrollHeight", {
            get: function () {
                return this._el.nativeElement.scrollHeight;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DirectiveStateService.prototype, "clientHeight", {
            get: function () {
                return this._el.nativeElement.clientHeight;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DirectiveStateService.prototype, "initMode", {
            get: function () {
                return this._initMode;
            },
            set: function (initMode) {
                this._initMode = initMode;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DirectiveStateService.prototype, "scrollStreamActive", {
            get: function () {
                return this._scrollStreamActive;
            },
            set: function (active) {
                this._scrollStreamActive = active;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DirectiveStateService.prototype, "previousScrollPositionpUpdated", {
            get: function () {
                return this._previousScrollPositionpUpdated;
            },
            set: function (previousScrollPositionpUpdated) {
                this._previousScrollPositionpUpdated = previousScrollPositionpUpdated;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DirectiveStateService.prototype, "previousScrollTop", {
            get: function () {
                return this._previousScrollTop;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(DirectiveStateService.prototype, "previousScrollHeight", {
            get: function () {
                return this._previousScrollHeight;
            },
            enumerable: false,
            configurable: true
        });
        DirectiveStateService.prototype.setup = function (params) {
            this._el = params.el;
            this._initMode = params.initMode;
            this._scrollStreamActive = params.scrollStreamActive;
            this._previousScrollPositionpUpdated = params.previousScrollPositionpUpdated;
            this.updatePreviousScrollTop();
            this.updatePreviousScrollHeight();
        };
        DirectiveStateService.prototype.updatePreviousScrollTop = function () {
            this._previousScrollTop = this._el.nativeElement.scrollTop;
        };
        DirectiveStateService.prototype.updatePreviousScrollHeight = function () {
            this._previousScrollHeight = this._el.nativeElement.scrollHeight;
        };
        return DirectiveStateService;
    }());
    DirectiveStateService.decorators = [
        { type: core.Injectable }
    ];

    var InitialScrollPosition;
    (function (InitialScrollPosition) {
        InitialScrollPosition["DEFAULT"] = "DEFAULT";
        InitialScrollPosition["TOP"] = "TOP";
        InitialScrollPosition["MIDDLE"] = "MIDDLE";
        InitialScrollPosition["BOTTOM"] = "BOTTOM";
    })(InitialScrollPosition || (InitialScrollPosition = {}));

    var DirectiveContext = /** @class */ (function () {
        function DirectiveContext() {
        }
        return DirectiveContext;
    }());

    var StrategyBase = /** @class */ (function () {
        function StrategyBase(directive, state) {
            this.directive = directive;
            this.state = state;
        }
        StrategyBase.prototype.wasScrolledDown = function (prevPos, currentPos) {
            return prevPos.scrollTop < currentPos.scrollTop;
        };
        StrategyBase.prototype.wasScrolledUp = function (prevPos, currentPos) {
            return !this.wasScrolledDown(prevPos, currentPos);
        };
        StrategyBase.prototype.isScrollDownEnough = function (pos, scrollPositionTrigger) {
            return ((pos.scrollTop + pos.clientHeight) / pos.scrollHeight) > (scrollPositionTrigger / 100);
        };
        StrategyBase.prototype.isScrollUpEnough = function (pos, scrollPositionTrigger) {
            return (pos.scrollTop / pos.scrollHeight) < (scrollPositionTrigger / 100);
        };
        StrategyBase.prototype.getInitialScrollPositionValue = function (defaultScrollPosition) {
            var initialScrollPosition = this.directive.initialScrollPosition;
            if (isNumber__default['default'](initialScrollPosition)) {
                return Number(initialScrollPosition);
            }
            var initialScrollPositions = this.getInitialScrollPositions();
            if (initialScrollPosition === InitialScrollPosition.DEFAULT) {
                return initialScrollPositions[defaultScrollPosition];
            }
            return initialScrollPositions[initialScrollPosition];
        };
        StrategyBase.prototype.getInitialScrollPositions = function () {
            var _a;
            var _b = this.state, scrollHeight = _b.scrollHeight, clientHeight = _b.clientHeight;
            return _a = {},
                _a[InitialScrollPosition.TOP] = 0,
                _a[InitialScrollPosition.MIDDLE] = scrollHeight / 2 - clientHeight / 2,
                _a[InitialScrollPosition.BOTTOM] = scrollHeight,
                _a;
        };
        return StrategyBase;
    }());

    var ScrollingToTop = /** @class */ (function (_super) {
        __extends(ScrollingToTop, _super);
        function ScrollingToTop(directive, state) {
            return _super.call(this, directive, state) || this;
        }
        ScrollingToTop.prototype.scrollDirectionChanged = function (scrollPairChanged) {
            var _this = this;
            return scrollPairChanged.pipe(operators.filter(function (scrollPositions) {
                return _super.prototype.wasScrolledUp.call(_this, scrollPositions[0], scrollPositions[1]);
            }));
        };
        ScrollingToTop.prototype.scrollRequestZoneChanged = function (scrollDirectionChanged) {
            var _this = this;
            return scrollDirectionChanged.pipe(operators.filter(function (scrollPositions) {
                return _super.prototype.isScrollUpEnough.call(_this, scrollPositions[1], _this.directive.scrollUpPercentilePositionTrigger);
            }));
        };
        ScrollingToTop.prototype.askForUpdate = function () {
            this.directive.onScrollUp.next();
        };
        ScrollingToTop.prototype.setInitialScrollPosition = function () {
            var initialScrollPositionValue = _super.prototype.getInitialScrollPositionValue.call(this, InitialScrollPosition.BOTTOM);
            this.directive.scrollTo(initialScrollPositionValue);
        };
        ScrollingToTop.prototype.setPreviousScrollPosition = function () {
            var prevScrollPosition = this.state.previousScrollTop +
                (this.state.scrollHeight - this.state.previousScrollHeight);
            this.directive.scrollTo(prevScrollPosition);
        };
        return ScrollingToTop;
    }(StrategyBase));

    var ScrollingToBottom = /** @class */ (function (_super) {
        __extends(ScrollingToBottom, _super);
        function ScrollingToBottom(directive, state) {
            return _super.call(this, directive, state) || this;
        }
        ScrollingToBottom.prototype.scrollDirectionChanged = function (scrollPairChanged) {
            var _this = this;
            return scrollPairChanged.pipe(operators.filter(function (scrollPositions) {
                return _super.prototype.wasScrolledDown.call(_this, scrollPositions[0], scrollPositions[1]);
            }));
        };
        ScrollingToBottom.prototype.scrollRequestZoneChanged = function (scrollDirectionChanged) {
            var _this = this;
            return scrollDirectionChanged.pipe(operators.filter(function (scrollPositions) {
                return _super.prototype.isScrollDownEnough.call(_this, scrollPositions[1], _this.directive.scrollDownPercentilePositionTrigger);
            }));
        };
        ScrollingToBottom.prototype.askForUpdate = function () {
            this.directive.onScrollDown.next();
        };
        ScrollingToBottom.prototype.setInitialScrollPosition = function () {
            var initialScrollPositionValue = _super.prototype.getInitialScrollPositionValue.call(this, InitialScrollPosition.TOP);
            this.directive.scrollTo(initialScrollPositionValue);
        };
        ScrollingToBottom.prototype.setPreviousScrollPosition = function () {
            var prevScrollPosition = this.state.previousScrollTop;
            this.directive.scrollTo(prevScrollPosition);
        };
        return ScrollingToBottom;
    }(StrategyBase));

    var ScrollingToBoth = /** @class */ (function (_super) {
        __extends(ScrollingToBoth, _super);
        function ScrollingToBoth(directive, state) {
            return _super.call(this, directive, state) || this;
        }
        ScrollingToBoth.prototype.scrollDirectionChanged = function (scrollPairChanged) {
            return scrollPairChanged;
        };
        ScrollingToBoth.prototype.scrollRequestZoneChanged = function (scrollDirectionChanged) {
            var _this = this;
            return scrollDirectionChanged.pipe(operators.filter(function (scrollPositions) {
                return (_super.prototype.isScrollUpEnough.call(_this, scrollPositions[1], _this.directive.scrollUpPercentilePositionTrigger) || _super.prototype.isScrollDownEnough.call(_this, scrollPositions[1], _this.directive.scrollDownPercentilePositionTrigger));
            }), operators.tap(function (scrollPositions) {
                _this.scrolledUp = _super.prototype.wasScrolledUp.call(_this, scrollPositions[0], scrollPositions[1]);
            }));
        };
        ScrollingToBoth.prototype.askForUpdate = function () {
            if (this.scrolledUp) {
                this.directive.onScrollUp.next();
            }
            else {
                this.directive.onScrollDown.next();
            }
        };
        ScrollingToBoth.prototype.setInitialScrollPosition = function () {
            var initialScrollPositionValue = _super.prototype.getInitialScrollPositionValue.call(this, InitialScrollPosition.MIDDLE);
            this.directive.scrollTo(initialScrollPositionValue);
        };
        ScrollingToBoth.prototype.setPreviousScrollPosition = function () {
            var prevScrollPosition;
            if (this.scrolledUp) {
                prevScrollPosition = this.state.previousScrollTop +
                    (this.state.scrollHeight - this.state.previousScrollHeight);
            }
            else {
                prevScrollPosition = this.state.previousScrollTop;
            }
            this.directive.scrollTo(prevScrollPosition);
        };
        return ScrollingToBoth;
    }(StrategyBase));

    var ScrollHeightListener = /** @class */ (function () {
        function ScrollHeightListener(directive, state) {
            this.directive = directive;
            this.state = state;
            this.DEFAULT_REQUEST_TIMEOUT = 30000;
        }
        ScrollHeightListener.prototype.start = function () {
            var _this = this;
            this.listener = window.requestAnimationFrame(this.listen.bind(this));
            if (!this.httpRequestTimeout) {
                this.httpRequestTimeout = setTimeout(function () {
                    _this.stopIfRequestTimeout();
                }, this.DEFAULT_REQUEST_TIMEOUT);
            }
        };
        ScrollHeightListener.prototype.stop = function () {
            window.cancelAnimationFrame(this.listener);
            clearTimeout(this.httpRequestTimeout);
            this.httpRequestTimeout = null;
        };
        ScrollHeightListener.prototype.listen = function () {
            if (this.state.previousScrollHeight !== this.state.scrollHeight) {
                this.stop();
                this.directive.onScrollbarHeightChanged();
            }
            else {
                this.start();
            }
        };
        ScrollHeightListener.prototype.stopIfRequestTimeout = function () {
            if (!this.state.previousScrollPositionpUpdated) {
                this.stop();
            }
        };
        return ScrollHeightListener;
    }());

    var NgxInfiniteScrollerDirective = /** @class */ (function (_super) {
        __extends(NgxInfiniteScrollerDirective, _super);
        function NgxInfiniteScrollerDirective(platformId, el, renderer, state) {
            var _this = _super.call(this) || this;
            _this.platformId = platformId;
            _this.el = el;
            _this.renderer = renderer;
            _this.state = state;
            _this.strategy = 'scrollingToBottom';
            _this.initialScrollPosition = InitialScrollPosition.DEFAULT;
            _this.scrollbarAnimationInterval = 100;
            _this.scrollDebounceTimeAfterScrollHeightChanged = 50;
            _this.scrollDebounceTimeAfterDOMMutationOnInit = 1000;
            _this.scrollUpPercentilePositionTrigger = 2;
            _this.scrollDownPercentilePositionTrigger = 98;
            _this.onScrollUp = new core.EventEmitter();
            _this.onScrollDown = new core.EventEmitter();
            _this.scrollHeightChanged = new rxjs.Subject();
            _this.domMutationEmitter = new rxjs.Subject();
            _this.isBrowser = common.isPlatformBrowser(platformId);
            _this.state.setup({
                el: el,
                initMode: true,
                scrollStreamActive: true,
                previousScrollPositionpUpdated: false
            });
            return _this;
        }
        Object.defineProperty(NgxInfiniteScrollerDirective.prototype, "scrollPairChanged", {
            get: function () {
                var _this = this;
                if (this.scrollChanged) {
                    return this.scrollChanged.pipe(operators.takeWhile(function () { return _this.state.scrollStreamActive; }), operators.map(function (e) {
                        return {
                            scrollHeight: e.target.scrollHeight,
                            scrollTop: e.target.scrollTop,
                            clientHeight: e.target.clientHeight,
                        };
                    }), operators.pairwise(), operators.debounceTime(this.scrollbarAnimationInterval));
                }
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(NgxInfiniteScrollerDirective.prototype, "scrollDirectionChanged", {
            get: function () {
                return this.scrollingStrategy.scrollDirectionChanged(this.scrollPairChanged);
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(NgxInfiniteScrollerDirective.prototype, "scrollRequestZoneChanged", {
            get: function () {
                var _this = this;
                return this.scrollingStrategy.scrollRequestZoneChanged(this.scrollDirectionChanged).pipe(operators.tap(function () {
                    _this.state.updatePreviousScrollTop();
                    _this.state.updatePreviousScrollHeight();
                    _this.state.previousScrollPositionpUpdated = false;
                    _this.scrollHeightListener.start();
                }));
            },
            enumerable: false,
            configurable: true
        });
        NgxInfiniteScrollerDirective.prototype.ngOnInit = function () {
            this.useStrategy();
            this.useScrollHeightListener();
            this.registerScrollEventHandler();
            this.registerMutationObserver();
            this.registerInitialScrollPostionHandler();
            this.registerPreviousScrollPositionHandler();
        };
        NgxInfiniteScrollerDirective.prototype.ngAfterViewInit = function () {
            this.registerScrollSpy();
        };
        NgxInfiniteScrollerDirective.prototype.ngOnDestroy = function () {
            this.unregisterMutationObserver();
        };
        NgxInfiniteScrollerDirective.prototype.scrollTo = function (position) {
            this.state.scrollStreamActive = false;
            this.renderer.setProperty(this.el.nativeElement, 'scrollTop', position);
            this.state.scrollStreamActive = true;
        };
        NgxInfiniteScrollerDirective.prototype.onScrollbarHeightChanged = function () {
            this.scrollHeightChanged.next();
        };
        NgxInfiniteScrollerDirective.prototype.registerScrollEventHandler = function () {
            this.scrollChanged = rxjs.fromEvent(this.el.nativeElement, 'scroll');
        };
        NgxInfiniteScrollerDirective.prototype.registerMutationObserver = function () {
            var _this = this;
            if (this.isBrowser) {
                this.domMutationObserver = new MutationObserver(function (mutations) {
                    _this.domMutationEmitter.next(mutations);
                });
                var config = { attributes: true, childList: true, characterData: true };
                this.domMutationObserver.observe(this.el.nativeElement, config);
            }
        };
        NgxInfiniteScrollerDirective.prototype.registerInitialScrollPostionHandler = function () {
            var _this = this;
            this.domMutationEmitter.pipe(operators.takeWhile(function () { return _this.state.initMode; }), operators.debounceTime(this.scrollDebounceTimeAfterDOMMutationOnInit)).subscribe(function () {
                _this.scrollingStrategy.setInitialScrollPosition();
                _this.state.initMode = false;
            });
        };
        NgxInfiniteScrollerDirective.prototype.registerPreviousScrollPositionHandler = function () {
            var _this = this;
            rxjs.zip(this.scrollRequestZoneChanged, this.scrollHeightChanged).pipe(operators.skipWhile(function () { return _this.state.initMode; }), operators.debounceTime(this.scrollDebounceTimeAfterScrollHeightChanged)).subscribe(function () {
                _this.scrollingStrategy.setPreviousScrollPosition();
                _this.state.previousScrollPositionpUpdated = true;
            });
        };
        NgxInfiniteScrollerDirective.prototype.registerScrollSpy = function () {
            var _this = this;
            this.scrollRequestZoneChanged.subscribe(function () {
                _this.scrollingStrategy.askForUpdate();
            });
        };
        NgxInfiniteScrollerDirective.prototype.unregisterMutationObserver = function () {
            if (this.domMutationObserver) {
                this.domMutationObserver.disconnect();
            }
        };
        NgxInfiniteScrollerDirective.prototype.useStrategy = function () {
            switch (this.strategy) {
                case 'scrollingToBoth':
                    this.scrollingStrategy = new ScrollingToBoth(this, this.state);
                    break;
                case 'scrollingToTop':
                    this.scrollingStrategy = new ScrollingToTop(this, this.state);
                    break;
                case 'scrollingToBottom':
                default:
                    this.scrollingStrategy = new ScrollingToBottom(this, this.state);
                    break;
            }
        };
        NgxInfiniteScrollerDirective.prototype.useScrollHeightListener = function () {
            this.scrollHeightListener = new ScrollHeightListener(this, this.state);
        };
        return NgxInfiniteScrollerDirective;
    }(DirectiveContext));
    NgxInfiniteScrollerDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[ngxInfiniteScroller]'
                },] }
    ];
    NgxInfiniteScrollerDirective.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
        { type: core.ElementRef },
        { type: core.Renderer2 },
        { type: DirectiveStateService }
    ]; };
    NgxInfiniteScrollerDirective.propDecorators = {
        strategy: [{ type: core.Input }],
        initialScrollPosition: [{ type: core.Input }],
        scrollbarAnimationInterval: [{ type: core.Input }],
        scrollDebounceTimeAfterScrollHeightChanged: [{ type: core.Input }],
        scrollDebounceTimeAfterDOMMutationOnInit: [{ type: core.Input }],
        scrollUpPercentilePositionTrigger: [{ type: core.Input }],
        scrollDownPercentilePositionTrigger: [{ type: core.Input }],
        onScrollUp: [{ type: core.Output }],
        onScrollDown: [{ type: core.Output }]
    };

    var NgxInfiniteScrollerModule = /** @class */ (function () {
        function NgxInfiniteScrollerModule() {
        }
        return NgxInfiniteScrollerModule;
    }());
    NgxInfiniteScrollerModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [
                        NgxInfiniteScrollerDirective
                    ],
                    imports: [],
                    exports: [
                        NgxInfiniteScrollerDirective
                    ],
                    providers: [
                        DirectiveStateService
                    ],
                    bootstrap: []
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NgxInfiniteScrollerModule = NgxInfiniteScrollerModule;
    exports.ɵa = NgxInfiniteScrollerDirective;
    exports.ɵb = DirectiveContext;
    exports.ɵc = DirectiveStateService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
