// ngx-infinite-scroller/esm2015/src/app/directive-state.service.js
import { Injectable } from '@angular/core';
export class DirectiveStateService {
    get scrollTop() {
        return this._el.nativeElement.scrollTop;
    }
    get scrollHeight() {
        return this._el.nativeElement.scrollHeight;
    }
    get clientHeight() {
        return this._el.nativeElement.clientHeight;
    }
    get initMode() {
        return this._initMode;
    }
    set initMode(initMode) {
        this._initMode = initMode;
    }
    get scrollStreamActive() {
        return this._scrollStreamActive;
    }
    set scrollStreamActive(active) {
        this._scrollStreamActive = active;
    }
    get previousScrollPositionpUpdated() {
        return this._previousScrollPositionpUpdated;
    }
    set previousScrollPositionpUpdated(previousScrollPositionpUpdated) {
        this._previousScrollPositionpUpdated = previousScrollPositionpUpdated;
    }
    get previousScrollTop() {
        return this._previousScrollTop;
    }
    get previousScrollHeight() {
        return this._previousScrollHeight;
    }
    setup(params) {
        this._el = params.el;
        this._initMode = params.initMode;
        this._scrollStreamActive = params.scrollStreamActive;
        this._previousScrollPositionpUpdated = params.previousScrollPositionpUpdated;
        this.updatePreviousScrollTop();
        this.updatePreviousScrollHeight();
    }
    updatePreviousScrollTop() {
        this._previousScrollTop = this._el.nativeElement.scrollTop;
    }
    updatePreviousScrollHeight() {
        this._previousScrollHeight = this._el.nativeElement.scrollHeight;
    }
}
DirectiveStateService.decorators = [
    { type: Injectable }
];
